File: typescript/tests/ambient_board.test.ts
// ambient_board.test.ts
// Verifies the CircleAI.Ambient port: latest/history newest-first, preferences,
// and the comfort test (temp/humidity tolerances + noise ceiling).

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryAmbientBoard,
  ambientReading,
  ambientPreference,
} from "../src/ambient/index";

describe("InMemoryAmbientBoard", () => {
  it("returns the latest reading and history newest-first capped by limit", () => {
    const b = new InMemoryAmbientBoard();
    b.record(ambientReading("d1", 22, 50, 300, 40, new Date("2026-01-01T08:00:00Z")));
    b.record(ambientReading("d1", 23, 48, 320, 42, new Date("2026-01-01T12:00:00Z")));
    b.record(ambientReading("d2", 30, 60, 100, 55, new Date("2026-01-01T12:00:00Z")));
    assert.equal(b.latest("d1")?.temperatureC, 23);
    assert.deepEqual(
      b.history("d1").map((r) => r.atUtc.toISOString()),
      ["2026-01-01T12:00:00.000Z", "2026-01-01T08:00:00.000Z"],
    );
    assert.equal(b.history("d1", 1).length, 1);
  });

  it("latest is undefined for an unseen device", () => {
    const b = new InMemoryAmbientBoard();
    assert.equal(b.latest("nope"), undefined);
  });

  it("comfort requires a preference and a reading within tolerances", () => {
    const b = new InMemoryAmbientBoard();
    assert.equal(b.isComfortable("d1", "office"), false); // no pref, no reading
    b.setPreference(ambientPreference("office", 22, 50, 45));
    assert.equal(b.getPreference("office")?.targetTempC, 22);
    assert.equal(b.isComfortable("d1", "office"), false); // pref but no reading

    // Within tolerance: |23-22|<=2, |55-50|<=10, 44<=45.
    b.record(ambientReading("d1", 23, 55, 300, 44, new Date("2026-01-01T12:00:00Z")));
    assert.equal(b.isComfortable("d1", "office"), true);
  });

  it("comfort fails when any bound is exceeded", () => {
    const b = new InMemoryAmbientBoard();
    b.setPreference(ambientPreference("office", 22, 50, 45));
    // Noise too high.
    b.record(ambientReading("d1", 22, 50, 300, 46, new Date("2026-01-01T12:00:00Z")));
    assert.equal(b.isComfortable("d1", "office"), false);
    // Temp too far off.
    b.record(ambientReading("d1", 26, 50, 300, 40, new Date("2026-01-01T13:00:00Z")));
    assert.equal(b.isComfortable("d1", "office"), false);
  });
});

File: go/tests/ambient_board_test.go
// ambient_board_test.go
//
// Verifies the CircleAI.Ambient port (ambient_board.go): reading record, latest
// + newest-first history with limit, preference set/get, and comfort evaluation
// against the temp/humidity/noise thresholds.

package circleai_test

import (
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestAmbient_ReadingsAndHistory(t *testing.T) {
	b := circleai.NewInMemoryAmbientBoard()
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	b.Record(circleai.AmbientReading{DeviceId: "d1", TemperatureC: 21, Humidity: 45, LuxLight: 300, DbNoise: 35, AtUtc: base})
	b.Record(circleai.AmbientReading{DeviceId: "d1", TemperatureC: 22, Humidity: 46, LuxLight: 320, DbNoise: 36, AtUtc: base.Add(time.Hour)})
	b.Record(circleai.AmbientReading{DeviceId: "d2", TemperatureC: 30, Humidity: 70, LuxLight: 100, DbNoise: 60, AtUtc: base})

	last, ok := b.Latest("d1")
	if !ok || last.TemperatureC != 22 {
		t.Fatalf("latest = %+v ok=%v", last, ok)
	}
	hist := b.History("d1", 50)
	if len(hist) != 2 || hist[0].TemperatureC != 22 || hist[1].TemperatureC != 21 {
		t.Fatalf("history newest-first failed: %+v", hist)
	}
	if lim := b.History("d1", 1); len(lim) != 1 || lim[0].TemperatureC != 22 {
		t.Fatalf("history limit failed: %+v", lim)
	}
}

func TestAmbient_Comfort(t *testing.T) {
	b := circleai.NewInMemoryAmbientBoard()
	now := time.Now().UTC()
	b.SetPreference(circleai.AmbientPreference{Location: "office", TargetTempC: 22, TargetHumidity: 45, MaxNoiseDb: 40})
	if got, ok := b.GetPreference("office"); !ok || got.TargetTempC != 22 {
		t.Fatalf("get preference = %+v ok=%v", got, ok)
	}

	// Within all thresholds -> comfortable.
	b.Record(circleai.AmbientReading{DeviceId: "d1", TemperatureC: 23, Humidity: 50, DbNoise: 38, AtUtc: now})
	if !b.IsComfortable("d1", "office") {
		t.Fatalf("expected comfortable")
	}
	// Too loud -> not comfortable.
	b.Record(circleai.AmbientReading{DeviceId: "d1", TemperatureC: 22, Humidity: 45, DbNoise: 55, AtUtc: now.Add(time.Minute)})
	if b.IsComfortable("d1", "office") {
		t.Fatalf("expected not comfortable (noise)")
	}
	// Missing preference or reading -> false.
	if b.IsComfortable("d1", "nowhere") {
		t.Fatalf("missing preference must be not comfortable")
	}
	if b.IsComfortable("ghost", "office") {
		t.Fatalf("missing reading must be not comfortable")
	}
}
